### PR TITLE
Migrate @woocommerce/data navigation store to TS

### LINKED ISSUE #33839: Migrate @woocommerce/data navigation store to TS
### Issue Description:

Migrate @woocommerce/data [navigation store](https://github.com/woocommerce/woocommerce/tree/trunk/packages/js/data/src/navigation) to TypeScript.

Example PR
- https://github.com/woocommerce/woocommerce/pull/33482

### Additional context
https://github.com/woocommerce/woocommerce/issues/32179

pdibGW-y2-p2

### PR DESCRIPTION
**Pull request by chihsuan**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #33839.

This PR migrates the navigation store to Typescript.

### How to test the changes in this Pull Request:

All GH action should pass

1. Use a fresh site
2. Install and activate [add-navigation-items.zip](https://github.com/woocommerce/woocommerce/files/9286552/add-navigation-items.zip) plugin 
3. Go to `Woocommerce > Setting > Advanced > Features` (http://localhost:8888/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features
4. Enable "Navigation" feature
5. Go to top menu level and click on `Example Category` menu item
6. Should see `Sub Menu Child 1` and `JavaScript Example` under `Example Category`
7. Click on the "star" icon
8. Go back to top menu level
9. Observe that `Example Category` is under "WooCommerce" section.
10. Please test around the navigation feature

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/data/src/navigation/actions.ts [hunk 1]
```diff
@@ -9,91 +9,87 @@ import { getPersistedQuery } from '@woocommerce/navigation';
...
+export function removeFavoriteFailure( favorite: string, error: unknown ) {
 	return {
 		type: TYPES.REMOVE_FAVORITE_FAILURE,
 		favorite,
 		error,
 	};
 }
 
-export function removeFavoriteSuccess( favorite, error ) {
+export function removeFavoriteSuccess( favorite: string ) {
```
Comment (ilyasfoo) [inline packages/js/data/src/navigation/actions.ts:86]: Nice catch 👍

File: packages/js/data/src/navigation/reducer.ts [hunk 2]
```diff
@@ -1,35 +1,43 @@
+/**
+ * External dependencies
+ */
+
```
Comment (ilyasfoo) [inline packages/js/data/src/navigation/reducer.ts]:
```suggestion
```
Extra space

File: packages/js/data/src/navigation/types.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * External dependencies
+ */
+import { getPersistedQuery } from '@woocommerce/navigation';
+
+export type MenuItem = {
+	id: string;
+	title: string;
+	url: string;
+	order: number;
+	migrate: boolean;
```
Comment (ilyasfoo) [inline packages/js/data/src/navigation/types.ts:12]: I wonder if this is should be a required parameter? I can't find any example uses though, would appreciate if you found any context

**Review by ilyasfoo — DISMISSED** [review]
This tested well, I have some nitpicky comments but otherwise LGTM! 🚢

File: packages/js/data/src/navigation/types.ts (hunk 3 above)
Comment (chihsuan) [reply packages/js/data/src/navigation/types.ts:12]: It should be a required parameter.  In general, I found the context in Menu.php, but it was hard to track by inspecting the code, so I actually checked the API response directly.

https://github.com/woocommerce/woocommerce/blob/ecec9eaa76cb7b2e36f79175837b71f4f64996b1/plugins/woocommerce/src/Admin/Features/Navigation/Menu.php#L159-L168

https://github.com/woocommerce/woocommerce/blob/ecec9eaa76cb7b2e36f79175837b71f4f64996b1/plugins/woocommerce/src/Admin/Features/Navigation/Menu.php#L207-L219

https://github.com/woocommerce/woocommerce/blob/ecec9eaa76cb7b2e36f79175837b71f4f64996b1/plugins/woocommerce/src/Admin/Features/Navigation/Menu.php#L356-L370
(menu_id is set inside the method)

Comment (chihsuan) [reply packages/js/data/src/navigation/types.ts:12]: I just added a helpful comment for the type in [02bdfa111570b980f3f40b2427278e1aa1982dfe](https://github.com/woocommerce/woocommerce/pull/34239/commits/02bdfa111570b980f3f40b2427278e1aa1982dfe). 🙌

Comment (ilyasfoo) [reply packages/js/data/src/navigation/types.ts:12]: Nice! I think that helps! Thank you

**Review by ilyasfoo — APPROVED** [review]
Nice work! LGTM 🚢

**Pull request merged**